### ISSUE TITLE
Several glossary entries about cluster architecture may lead to confusion

### ISSUE THREAD

**Issue by TransitivityPaul** [kubernetes/website#48556]
I'm not sure if this is the right channel, but..

For various reasons I'm modeling Kubernetes and am trying to land on definitions that are a little more precise so that I can reason about the structure.  Any thoughts below are as much to get a response around their correctness (so that I get my model right) as they are about suggested updates to the glossary itself.

Regarding the definition of **Cluster** - The terms "worker machine"  and "worker node" seem perhaps unnecessary and potentially confusing.  Especially as worker * may also (optionally) host elements of the control plane.  Is the distinction between a worker node and a non-worker node meaningful?  Is it possible to simplify the definition by saying that:

> A Kubernetes cluster is composed of a set of one or more nodes that are exclusively managed by an instance of the Kubernetes control plane.  The control plane runs on one of more of the nodes in the cluster, and is responsible for managing the cluster, and the containerized workloads (in the form of pods) that run on member nodes.  

As it stands, the definition also doesn't preclude the control plane being hosted outside of the cluster, even though the architecture diagram shows it being within the cluster.  I know of some cluster frameworks that do support "externalized" control planes or control plane components.  

**Control Plane** - It might make sense to reorder the words or to add a comma to make it clear that..

> The control plane provides API/Interface driven management of the lifecycles and service levels (performance, scale, availability etc,) of the containerized workloads that run on the cluster, and of the cluster itself.  Components of the control plane may be distributed across multiple nodes for scale and/or resilience.  Components of the control plane may also be co-located with containerized workloads, i.e. pods, on the same node.

**Node** could perhaps likewise be simplified...

> A node is a member of a Kubernetes cluster.  It consists of a machine, either a physical computer or a virtual machine, together with an operating system, and appropriate Kubernetes services.  A node may host control plane components, containerized workloads, in the form of Pods, or both...

Is there a reason to continue to refer to **Masters** and Minions at all?   If they still occur in documents that are not yet deprecated, then it might make sense to call this out specifically as glossary items.

> Master - A deprecated term, that is still extant in some older documents, referring to nodes dedicated to running the control plane.

 Otherwise it may make sense to drop the terms completely.  The term Master is somewhat pejorative, especially considering cluster role definitions of the past.  

Anyway, I hope you can help me me land on the accurate definitions for myself, and perhaps these may also help evolve the glossary.

Best regards
Paul

**Comment by sftim** [comment]
> Regarding the definition of **Cluster** - The terms "worker machine" and "worker node" seem perhaps unnecessary and potentially confusing. Especially as worker * may also (optionally) host elements of the control plane. Is the distinction between a worker node and a non-worker node meaningful? Is it possible to simplify the definition by saying that:
> 
> > A Kubernetes cluster is composed of a set of one or more nodes that are exclusively managed by an instance of the Kubernetes control plane.  The control plane runs on one of more of the nodes in the cluster, and is responsible for managing the cluster, and the containerized workloads (in the form of pods) that run on member nodes.

No; that wouldn't be correct. The control plane is always part of the cluster, but doesn't have to run as containers.

**Comment by sftim** [comment]
> As it stands, the definition also doesn't preclude the control plane being hosted outside of the cluster, even though the architecture diagram shows it being within the cluster. I know of some cluster frameworks that do support "externalized" control planes or control plane components.

The control plane is always part of the cluster, but you don't have to use Kubernetes technology to supervise and deploy it.

**Comment by sftim** [comment]
> **Node** could perhaps likewise be simplified...
> 
> > A node is a member of a Kubernetes cluster.  It consists of a machine, either a physical computer or a virtual machine, together with an operating system, and appropriate Kubernetes services.  A node may host control plane components, containerized workloads, in the form of Pods, or both...

You are welcome to send in a pull request.
 
> Is there a reason to continue to refer to **Masters** and Minions at all? If they still occur in documents that are not yet deprecated, then it might make sense to call this out specifically as glossary items.

Yes, lots of documentation and code in the wider ecosystem still uses the term _master_. Some very old docs talk about minions.

We could move the minion explanation to its own glossary entry; contributions are, again, very welcome.

/language en

**Comment by sftim** [comment]
> Is the distinction between a worker node and a non-worker node meaningful?

A worker node is a node that isn't used to host any part of the control plane. If this isn't clear, let's try to improve the docs we have.

**Comment by babugeet** [comment]
@sftim : can i take it up?
/assign

**Comment by TransitivityPaul** [comment]
Hi,
Firstly thanks for responding.
Secondly is this the best place for discussion? Please feel free to redirect me if not.
Finally - I guess my question about the distinction is less about the definition itself, but more about:
 - Whether a node without any part of the control plane is treated fundamentally differently from other nodes in any way?  Does it get preferential treatment or otherwise?
 - Whether you can have a cluster with no worker nodes, i.e. that each node hosts some aspect of the control plane?  For example a basic 2 node cluster with that has control plane elements on both nodes?
Again thanks for engaging, and helping me contribute as best I can.
Cheers
Paul

**Comment by sftim** [comment]
Best place for discussion? Likely https://discuss.kubernetes.io/

> Whether a node without any part of the control plane is treated fundamentally differently from other nodes in any way?  Does it get preferential treatment or otherwise?

I'd take that question to the forums. This repo is for improving the docs.

> Whether you can have a cluster with no worker nodes, i.e. that each node hosts some aspect of the control plane?  For example a basic 2 node cluster with that has control plane elements on both nodes?

You can have no nodes at all, but it's rarely done. The jargon term is _scaling to zero_. You also get clusters with only control plane nodes, and workloads scheduled onto the nodes that run the control plane. This is more common but still not the typical way to operate Kubernetes.

**Comment by sftim** [comment]
@TransitivityPaul if your main aim was to learn Kubernetes, please consider closing this issue.

**Comment by TransitivityPaul** [comment]
Hi Tim,

Not my aim.  But a fair question to ask.  

My aim is to improve the docs, and in so doing ensure that I have the right definitions for my needs.  I’d rather align with the place where everyone goes, rather than use terminology that is divergent and then have to say why my definition is slightly different.

For the work I do I need simplicity and precision.  And glossaries are where I look for the simple and precise canonical definitions, as it should be easier than looking through the code.  My point with worker node/machine, is that sometimes there is a difference without distinction.  Yes, descriptively there is a difference between a node that doesn’t host control plane components, or that doesn’t host non-control plane workloads.  But are they managed differently in any way by the control plane, operators or users?  If the answer to such questions as the latter is no, then my tendency is to deprecate the use of a term so that all future docs are simpler and easier to understand, and reference that deprecation to cover for where it hasn’t been removed from the broader corpus of docs.  Otherwise you’re left with lots of clutter - nomenclature debt if you will ;) - that can be confusing, and have interesting consequences...

...in my prior roles I have seen disastrous outcomes, in one case costing many, many millions of dollars, that resulted from the slightest of misunderstandings about the definition of a term describing a thing.  In this case the same, very common, term was used by two different teams to describe two things that were almost completely, but not quite, identical.  But when a reader assumed that the behavior of both was the same in a particular circumstance, it resulted in an extraordinarily expensive problem that impacted many millions of users.  These kinds of experiences have rather conditioned me to be perhaps a little obsessive when it comes to clarity, and precision wrt definitions ;) 

Anyway, a non goal of mine is to needlessly distract others.  Suffice to say I think what you do is important.  I’ll close the issue, and submit changes for those things that you’ve suggested.  And then I’ll look at the code, consult the community or whatever to see if the proletariat nodes are distinct.

Anon
Paul

**Comment by sftim** [comment]
/retitle Several glossary entries about cluster architecture may lead to confusion

It's more that the glossary entries fail to head off (avoid) the confusion, but we like terse issue titles.

We don't set out for the glossary to supersede or substitute for reading the actual concept guide, but it's plausible that a minority of readers will do the same thing.

**Comment by sftim** [comment]
See https://github.com/kubernetes/website/pull/48686
